## Problem Set 2: Merging and regular expressions

**Total points (without extra credit)**: 30 

**Background on the policy context**: here, we're going to use two datasets to practice reshaping, merging, and regular expression patterns. Both datasets relate to the broader issue of which employers might be violating the rights of temporary guestworkers granted visas under the H-2A program. Here are some articles about potential exploitation of guestworkers by firms and inequality caused by minimal oversight:

- News media coverage of labor abuses of temporary guestworkers: https://www.buzzfeednews.com/article/kenbensinger/the-pushovers 
- GAO report on labor abuses of temporary guestworkers: https://www.gao.gov/products/gao-15-154

The following datasets are located in `pset2_inputdata` (need to unzip): 

- `jobs_clean`: a dataset of guestworker jobs posted by many employers, some of whom have been debarred (banned) from the program for labor abuses; others not debarred
- `debar`: a dataset of employers who committed violations of labor regulations meant to protect temporary guestworkers 


You can view a codebook here: https://docs.google.com/spreadsheets/d/1rF9GJEC8pPKxipD0TsoG9DVdqz3EJ-b-BHEtyioAX7I/edit?usp=sharing


In [111]:
## helpful packages
import pandas as pd
import numpy as np
import random
import re
import os

## repeated printouts
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"


# 1. Reshaping data (13 points total)

Load the following dataset stored in `pset2_inputdata`: `debar.csv`

This represents employers temporarily banned from hiring workers (debar.csv); call this `debar`


View the head()


In [112]:
## your code here
debar = pd.read_csv("pset2_inputdata/debar.csv")
debar.head()

Name     City, State  \
0                    J&J Harvesting       Leads, ND   
1            Stahlman Apiaries, Inc       Selby, SD   
2                     Trust Nursery     Pulaski, NY   
3             Anton Fertilizer Inc.     Dighton, KS   
4  Great Plains Fluid Service, Inc.  Greensburg, KS   

                                        Violation Duration Start date  \
0  Failure to respond to audit (partial response)  2 years  1/19/2014   
1  Failure to respond to audit (partial response)   1 year  2/19/2015   
2  Failure to respond to audit (partial response)   1 year  3/21/2014   
3       Failure to respond to audit (no response)  2 years  3/30/2014   
4       Failure to respond to audit (no response)  2 years  3/30/2014   

    End date  
0  1/18/2016  
1  2/14/2016  
2  3/20/2015  
3  3/29/2016  
4  3/29/2016

## 1.1 (1 point)

Print the number of rows in `debar` versus the number of unique employer names (`Name`). Is there one row per employer or multiple rows for some employers?

In [113]:
## your code here
debar.shape[0]   
debar["Name"].nunique()

114

98

## Rows vs Employers
There are multiple rows for some employers.  There are 114 rows but only 98 unique employers.

## 1.2 Investigating duplicated rows (2 points)

A. Create a new column in `debar`--`is_repeated`-- that tells us whether an employer (`Name`) is repeated > 1 times

*Hint*: there are multiple ways to solve this but some possibilities to get the list of names that are repeated are:
- Using value_counts() on the `Name` variable and extracting the index from that value counts 
- Using groupby to count the rows attached to one name

B. Print the rows where `is_repeated == True` and interpret

C. Subset to the rows where `is_repeated == True` and save that data as `mult_debar`. Print the head() and shape

In [115]:
## your code here
repeated = debar.Name.value_counts()
mult = []
for x in repeated.index:
    if repeated[x] > 1:
        mult.append(x)
        
# create is_repeated column
debar["is_repeated"] = debar["Name"].isin(mult)

# print where is_repeated == True
debar[debar["is_repeated"]]

# subset
mult_debar = debar[debar["is_repeated"]]
mult_debar.head()
mult_debar.shape


Name       City, State  \
6                    Annabella Land & Cattle     Annabella, UT   
7                        Autumn Hill Orchard        Groton, MA   
8      Caddo Creek Ranch, dba Paradise Ranch         Caddo, TX   
11                     Loewen Harvesting LLC   Brownsville, TX   
12               Rollo Farm Labor Contractor         Miami, FL   
14                             Sharon Mathis        Tifton, GA   
15                                 SRT Farms        Morton, TX   
16                               Mark Duncan     Roosevelt, UT   
17          Maple Ridge Custom Services, LLC     Altheimer, AK   
18                                 F&W Farms       Ingalls, KS   
19                        Cisco Produce Inc.         Cairo, GA   
21   Old Tree Farms/Verpaalen Custom Service         Volga, SD   
24               Rollo Farm Labor Contractor         Miami, FL   
25                     Loewen Harvesting LLC    Brownfield, TX   
28     Caddo Creek Ranch, dba Paradise Ranch      Caddo, Texas   
29                       Autumn Hill Orchard        Groton, MA   
30                   Annabella Land & Cattle   Annabella, Utah   
31                      Macky and Brad Farms        Plains, TX   
51   Old Tree Farms/Verpaalen Custom Service         Volga, SD   
55                      Macky and Brad Farms        Plains, TX   
56                        Cisco Produce Inc.         Cairo, GA   
58                                 F&W Farms       Ingalls, KS   
59          Maple Ridge Custom Services, LLC     Altheimer, AR   
60                               Mark Duncan     Roosevelt, UT   
61                                 SRT Farms        Morton, TX   
62                             Sharon Mathis        Tifton, GA   
73                              Xavier Horne    Lyons, Georgia   
89                              Xavier Horne    Lyons, Georgia   
103                         Dove Creek Farms  Mount Vernon, TX   
106                             Turner Farms         Healy, KS   
109                         Dove Creek Farms  Mount Vernon, TX   
111                             Turner Farms         Healy, KS   

                                             Violation  Duration  Start date  \
6                                          Non Payment    1 year    5/9/2014   
7            Failure to respond to audit (no response)   2 years    7/6/2014   
8       Failure to respond to audit (partial response)   2 years   7/20/2014   
11      Failure to respond to audit (partial response)    1 year   8/20/2014   
12           Failure to respond to audit (no response)   2 years   8/23/2014   
14           Failure to respond to audit (no response)   2 years  11/16/2014   
15           Failure to respond to audit (no response)   2 years  11/16/2014   
16           Failure to respond to audit (no response)   2 years  11/16/2014   
17      Failure to respond to audit (partial response)   2 years  11/16/2014   
18      Failure to respond to audit (partial response)   2 years  12/10/2014   
19           Failure to respond to audit (no response)   2 years  12/10/2014   
21                                       WHD Debarment   3 years  12/11/2014   
24          Impeding the Audit Process – Non- Response   2 years   8/23/2014   
25      Impeding the Audit Process – Partial- Response    1 year   8/20/2014   
28      Impeding the Audit Process – Partial- Response   2 years   7/20/2014   
29          Impeding the Audit Process – Non- Response   2 years    7/6/2014   
30                                         Non-payment    1 year    5/9/2014   
31           Failure to respond to audit (no response)    1 year   2/13/2015   
51                                 Wage Hour Debarment   3 years   12/1/2014   
55      Impeding the Audit Process – Partial- Response    1 year   2/13/2015   
56          Impeding the Audit Process – Non- Response   2 years  12/10/2015   
58      Impeding the Audit Process – Partial- Response    1 year  12/10/2014   
59      Impeding the 

Name      City, State  \
6                 Annabella Land & Cattle    Annabella, UT   
7                     Autumn Hill Orchard       Groton, MA   
8   Caddo Creek Ranch, dba Paradise Ranch        Caddo, TX   
11                  Loewen Harvesting LLC  Brownsville, TX   
12            Rollo Farm Labor Contractor        Miami, FL   

                                         Violation Duration Start date  \
6                                      Non Payment   1 year   5/9/2014   
7        Failure to respond to audit (no response)  2 years   7/6/2014   
8   Failure to respond to audit (partial response)  2 years  7/20/2014   
11  Failure to respond to audit (partial response)   1 year  8/20/2014   
12       Failure to respond to audit (no response)  2 years  8/23/2014   

     End date  is_repeated  
6    5/9/2015         True  
7    7/5/2016         True  
8   7/19/2016         True  
11  8/19/2015         True  
12  8/22/2016         True

(32, 7)

## 1.3 Reshape mult_debar to wide to begin filtering out duplicates (4 points)

You want to separate out two cases:

- Cases where the repeat rows for one employer are due to duplicated data 
- Cases where the repeat rows for one employer represent repeated violations for different issues

There are various ways to check duplicates in this data (eg converting `Violation` to lowercase; replacing spelled-out states with two-dig state codes)

We're going to use the simple rule of:

- A row is a duplicate if, within an employer (defined by Name + City, State), the Start date for each row's violation is the same 

To begin to check this, reshape `mult_debar` to a wide dataframe (`mult_debar_wide`) with the following columns, treating the `Name` and `City, State` as the index for the pivot:

- Name
- City, State
- start_date_viol1
- start_date_viol2

Print the head and shape

In [116]:
## your code here
mult_debar_new = mult_debar.copy()
state = ["Annabella, Utah", "Caddo, Texas", "Brownsville, TX", "Altheimer, AK"]
replace_state = ["Annabella, UT", "Caddo, TX", "Brownfield, TX", "Altheimer, AR"]
mult_debar_new["City, State"] = mult_debar_new["City, State"].replace(state, replace_state)

mult_debar_new.loc[:,"viol_num"] = mult_debar_new.groupby(["Name", "City, State"]).cumcount() + 1
mult_debar_wide = mult_debar_new.pivot_table(index=["Name", "City, State"], columns="viol_num", values="Start date", aggfunc="first")
mult_debar_wide.columns = ["start_date_viol1", "start_date_viol2"]
mult_debar_wide.head()
mult_debar_wide.shape


start_date_viol1  \
Name                                  City, State                         
Annabella Land & Cattle               Annabella, UT            5/9/2014   
Autumn Hill Orchard                   Groton, MA               7/6/2014   
Caddo Creek Ranch, dba Paradise Ranch Caddo, TX               7/20/2014   
Cisco Produce Inc.                    Cairo, GA              12/10/2014   
Dove Creek Farms                      Mount Vernon, TX         2/9/2018   

                                                       start_date_viol2  
Name                                  City, State                        
Annabella Land & Cattle               Annabella, UT            5/9/2014  
Autumn Hill Orchard                   Groton, MA               7/6/2014  
Caddo Creek Ranch, dba Paradise Ranch Caddo, TX               7/20/2014  
Cisco Produce Inc.                    Cairo, GA              12/10/2015  
Dove Creek Farms                      Mount Vernon, TX         2/9/2018

(16, 2)

## 1.4 Filter out duplicates from original debar data (6 points)

A. Using `mult_debar_wide`, add a column `is_dup` that takes value of True for cases where start_date_viol1 == start_date_viol2 marking the row as a duplicate

B. Going back to the original long-format data you loaded at the beginning- `debar`
    - For employers where `is_dup == True` as indicated by your wide-format dataframe, only keep `violnum == viol1`
    - For all other employers (so is_dup == False and ones we didnt need to check duplicates for), keep all violnum
    - Remove the `is_repeated` column from the `debar` data

**Hint**: you can complete part B without a for loop; `pd.concat` with axis = 0 (row binding) is one way

Call the resulting dataframe `debar_clean` and print the shape and # of unique employer names

In [121]:
## your code here
# create is_dup column
mult_debar_wide["is_dup"] = mult_debar_wide["start_date_viol1"] == mult_debar_wide["start_date_viol2"]

# create list of duplicates
mult_debar_wide = mult_debar_wide.reset_index().set_index("Name")
dups = mult_debar_wide[mult_debar_wide["is_dup"]].index

# create viol_num column in debar to use when determining which rows to keep
debar.loc[:,"viol_num"] = debar.groupby(["Name", "City, State"]).cumcount() + 1

# keep the first violation of duplicates
dups_keep = debar[(debar["Name"].isin(dups)) & (debar["viol_num"] == 1)]

# keep the non duplicates
nondups_keep = debar[~debar["Name"].isin(dups)]

# merge the rows you want to keep
debar_clean = pd.concat([dups_keep, nondups_keep], axis = 0)

# remove the is_repeated row
debar = debar.drop(columns="is_repeated", errors="ignore")
debar = debar.drop(columns="viol_num")
debar_clean = debar_clean.drop(columns="viol_num", errors="ignore")

debar_clean

# print num rows of the df and the number of unique names
debar_clean.shape
debar_clean["Name"].nunique()

Name       City, State  \
6                              Annabella Land & Cattle     Annabella, UT   
7                                  Autumn Hill Orchard        Groton, MA   
8                Caddo Creek Ranch, dba Paradise Ranch         Caddo, TX   
11                               Loewen Harvesting LLC   Brownsville, TX   
12                         Rollo Farm Labor Contractor         Miami, FL   
..                                                 ...               ...   
107                                       Walker Place      Danville, IL   
108  County Fair Farm (company) and Andrew Williams...     Jefferson, ME   
110                                      Jesus Ledesma      Mulberry, FL   
112        B & R Harvesting and Paul Cruz (individual)  Collins, Georgia   
113                                        Delia Rojas    Lyons, Georgia   

                                             Violation  Duration  Start date  \
6                                          Non Payment    1 year    5/9/2014   
7            Failure to respond to audit (no response)   2 years    7/6/2014   
8       Failure to respond to audit (partial response)   2 years   7/20/2014   
11      Failure to respond to audit (partial response)    1 year   8/20/2014   
12           Failure to respond to audit (no response)   2 years   8/23/2014   
..                                                 ...       ...         ...   
107  Failure to comply with the employer's obligati...  2 months  11/19/2019   
108                                      WHD Debarment   3 years    3/8/2017   
110                Failure to Respond to Audit Request   2 years      2/8/18   
112                                      WHD Debarment   3 years      4/9/17   
113                                      WHD Debarment   3 years     4/14/17   

      End date  
6     5/9/2015  
7     7/5/2016  
8    7/19/2016  
11   8/19/2015  
12   8/22/2016  
..         ...  
107  1/26/2020  
108   3/8/2020  
110     2/8/20  
112     4/9/20  
113    4/14/20  

[105 rows x 6 columns]

(105, 6)

98

# 2. Merging and regex (17 points total)



## 2.1 Load data on job postings

The previous dataset contains a small subset of employers who faced temporary bans due to violations of H-2A program regulations

Since most of the bans have expired, we're going to see which of those employers posted new H-2A jobs in the first quarter of 2021 

Loading the `jobs_clean.csv` data stored in `pset2_inputdata`

In [122]:
# your code here to load the data 

jobs = pd.read_csv("pset2_inputdata/jobs.csv")


##  2.2 Try inner join on employer name  (2 points)

- Use the `EMPLOYER_NAME` field of the `jobs` dataset
- Use the `Name` field of the `debar_clean` dataset 

A. Use pd.merge with an inner join on those fields to see whether there are any exact matches. 

B. If there are exact matches, print the row(s) with exact matches



In [123]:
## your code here
merged_jobs = jobs.merge(debar_clean, left_on = "EMPLOYER_NAME", right_on = "Name", how = "inner")

merged_jobs

CASE_NUMBER                           CASE_STATUS  \
0  H-300-20287-876656  Determination Issued - Certification   

             RECEIVED_DATE            DECISION_DATE  \
0  2020-10-20 09:20:32.010  2020-11-09 00:00:00.000   

  TYPE_OF_EMPLOYER_APPLICATION H2A_LABOR_CONTRACTOR NATURE_OF_TEMPORARY_NEED  \
0          Individual Employer                    Y                 Seasonal   

  EMERGENCY_FILING   EMPLOYER_NAME TRADE_NAME_DBA  ... PHONE_TO_APPLY  \
0                Y  Rafael Barajas            NaN  ...    18632732686   

  EMAIL_TO_APPLY               WEBSITE_TO_APPLY TOTAL_ADDENDUM_A_RECORDS  \
0            NaN  https://seasonaljobs.dol.gov/                        7   

             Name       City, State                         Violation  \
0  Rafael Barajas  Sebring, Florida  Non-payment of certification fee   

   Duration  Start date   End date  
0    1 year   9/23/2016  9/22/2017  

[1 rows x 144 columns]

## 2.3 Targeted regex (10 points total)

You want to see if you can increase the exact match rate with some basic cleaning of each 
of the employer name fields in each dataset 

### 2.3.1 Converting to upper (2 points)

A. Convert the `EMPLOYER_NAME` and `Name` fields to uppercase using list comprehension rather than df.varname.str.upper() (it's fine to do a separate list comprehension line for each of the two columns)

B. Print a random sample of 15 values of each result

C. Assign the full vector of uppercase names back to the original data, writing over the original `EMPLOYER_NAME` and `Name` columns 


In [124]:
## insert your code to turn into uppercase here
jobs_upper = [name.upper() for name in jobs['EMPLOYER_NAME']]
debar_upper = [name.upper() for name in debar_clean['Name']]


In [125]:
## insert your code for the random sample
random.sample(jobs_upper, 15)
random.sample(debar_upper, 15)

['GERALD PORTER OPERATING LLC',
 'WALTHER FARMS LLC - TX',
 'PHILLIP W. HEBERT',
 'BONNIE PLANTS, INC.',
 'FARMERS ORGANICS',
 'CYPRESS CREEK FARMS, LLC',
 'BLUE MOUNTAIN CATTLE INC.',
 'FREYOU FARMS',
 'MELISSA FEEDERS LLC',
 'RYAN BURRIS',
 'ELMORE TRUCK REPAIR',
 'MCF4 SOLUTIONS, LLC',
 'STROHAUER FARMS INC.',
 'BURTON CHRISTMAS TREES LLC',
 'LINER SOURCE, INC.  ']

['VALLEY VIEW ORCHARDS',
 'TRINITY LIVESTOCK PARTNERSHIP',
 'DAVID C. MARTINEZ',
 'DOVE CREEK FARMS',
 'MAPLE RIDGE CUSTOM SERVICES, LLC',
 'AGECY II, LLC',
 'JIM AND ANN SHIPLEY WILLIAM SHIPLEY*',
 'J&J HARVESTING',
 'JOHN R. COOK',
 'LABATTE FARMS',
 'GERONIMO SHEEP CO.',
 'CISCO PRODUCE INC.',
 'OLSON FARMS',
 '69 FARMS LLC',
 'RAFAEL BARAJAS']

In [126]:
## insert your code for assigning the uppercase names back to the data
jobs['EMPLOYER_NAME'] = jobs_upper
debar_clean['Name'] = debar_upper

jobs.head()
debar_clean.head()

CASE_NUMBER                           CASE_STATUS  \
0  H-300-20199-721302      Determination Issued - Withdrawn   
1  H-300-20231-773906  Determination Issued - Certification   
2  H-300-20231-774123  Determination Issued - Certification   
3  H-300-20231-774151  Determination Issued - Certification   
4  H-300-20231-774508  Determination Issued - Certification   

             RECEIVED_DATE            DECISION_DATE  \
0  2020-07-17 14:50:40.840  2020-10-01 00:00:00.000   
1  2020-08-20 10:38:15.620  2020-10-01 00:00:00.000   
2  2020-08-24 15:33:14.340  2020-10-01 00:00:00.000   
3  2020-08-21 12:08:09.760  2020-10-01 00:00:00.000   
4  2020-08-20 10:17:34.530  2020-10-01 00:00:00.000   

  TYPE_OF_EMPLOYER_APPLICATION H2A_LABOR_CONTRACTOR NATURE_OF_TEMPORARY_NEED  \
0          Individual Employer                    N                 Seasonal   
1          Association - Agent                    N                 Seasonal   
2          Individual Employer                    N                 Seasonal   
3          Individual Employer                    N                 Seasonal   
4          Individual Employer                    Y                 Seasonal   

  EMERGENCY_FILING                       EMPLOYER_NAME  \
0                Y  FAZIO FARMS OPERATING COMPANY, LLC   
1                N                  CHARLIE SUNDERLAND   
2                N                   MICHAEL RUDEBUSCH   
3                N                        LODAHL FARMS   
4                N             DUNSON HARVESTING, INC.   

                TRADE_NAME_DBA  ... ADDENDUM_B_HOUSING_ATTACHED  \
0                          NaN  ...                           N   
1  Panter & Sunderland Nursery  ...                           N   
2                          NaN  ...                           N   
3                          NaN  ...                           Y   
4      Dunson Harvesting, Inc.  ...                           Y   

  TOTAL_HOUSING_RECORDS MEALS_PROVIDED MEALS_CHARGED  \
0                     1              Y         12.68   
1                     1              N           NaN   
2                     1              N           NaN   
3                     2              N           NaN   
4                     8              N           NaN   

  MEAL_REIMBURSEMENT_MINIMUM MEAL_REIMBURSEMENT_MAXIMUM PHONE_TO_APPLY  \
0                      12.68                       55.0    13607017661   
1                      12.68                       55.0    19318083783   
2                      12.68                       55.0    19369333827   
3                      12.68                       55.0    14069637560   
4                      12.68                       55.0    18632939888   

                  EMAIL_TO_APPLY                             WEBSITE_TO_APPLY  \
0           faziofarms@gmail.com                                          NaN   
1                            NaN  https://www.jobs4tn.gov/vosnet/Default.aspx   
2  fayethlynpitre@rocketmail.com                                          NaN   
3        lodahl_kelsey@yahoo.com                                          NaN   
4                            NaN                        www.employflorida.com   

   TOTAL_ADDENDUM_A_RECORDS  
0                         0  
1                         0  
2                         0  
3                         0  
4                         4  

[5 rows x 138 columns]

Name      City, State  \
6                 ANNABELLA LAND & CATTLE    Annabella, UT   
7                     AUTUMN HILL ORCHARD       Groton, MA   
8   CADDO CREEK RANCH, DBA PARADISE RANCH        Caddo, TX   
11                  LOEWEN HARVESTING LLC  Brownsville, TX   
12            ROLLO FARM LABOR CONTRACTOR        Miami, FL   

                                         Violation Duration Start date  \
6                                      Non Payment   1 year   5/9/2014   
7        Failure to respond to audit (no response)  2 years   7/6/2014   
8   Failure to respond to audit (partial response)  2 years  7/20/2014   
11  Failure to respond to audit (partial response)   1 year  8/20/2014   
12       Failure to respond to audit (no response)  2 years  8/23/2014   

     End date  
6    5/9/2015  
7    7/5/2016  
8   7/19/2016  
11  8/19/2015  
12  8/22/2016

### 2.3.2 Cleaning up punctuation (4 points)

You notice that INC, CO, and LLC are sometimes followed by a period (.) but sometimes not

A. For each dataset, write a regex pattern using `re.sub` to remove the . but only if it's preceded by INC, LLC, or CO 

Make sure LLC, INC, CO remain part of the string but just without the dot

B. Test the pattern on the positive and negative example we provide below and print the result. See the Github issue for examples of what to return


**Hint**: https://stackoverflow.com/questions/7191209/python-re-sub-replace-with-matched-content



In [127]:
pos_example_1 = "CISCO PRODUCE INC."
pos_example_2 = "AVOYELLES HONEY CO., LLC"
neg_example = "E.V. RANCH LLP"

In [128]:
## insert your code here with the regex pattern for part A
regex = r"(INC|LLC|CO)\."

## insert your code to use re.sub to apply the pattern to the test cases for part B
res = [re.sub(regex, r"\1", name) for name in [pos_example_1, pos_example_2, neg_example]]
res

['CISCO PRODUCE INC', 'AVOYELLES HONEY CO, LLC', 'E.V. RANCH LLP']

### 2.3.3 (4 points)

Use that pattern in conjunction with `re.sub` and list comprehension to clean the employer name columns in each dataset. Save the new columns as `name_clean` in each. Then, use row subsetting to (1) subset to rows that changed names and (2) for:

- `debar_clean` print the `Name` and `name_clean` columns
- `jobs` print the `EMPLOYER_NAME` and `name_clean` columns

Make sure to use the uppercase versions of the variables


In [130]:
## your code here to clean the columns
# remove the periods
jobs_punc = [re.sub(regex, r"\1", name) for name in jobs['EMPLOYER_NAME']]
debar_punc = [re.sub(regex, r"\1", name) for name in debar_clean['Name']]

# assigne the new columns
jobs["name_clean"] = jobs_punc
debar_clean["name_clean"] = debar_punc

# create the subsetted dataframes that only include changed rows
sub_jobs = jobs[jobs["name_clean"] != jobs["EMPLOYER_NAME"]]
sub_debar = debar_clean[debar_clean["name_clean"] != debar_clean["Name"]]
sub_jobs[["EMPLOYER_NAME", "name_clean"]]
sub_debar[["Name", "name_clean"]]


EMPLOYER_NAME  \
4                      DUNSON HARVESTING, INC.   
7     FARM LABOR ASSOCIATION FOR GROWERS, INC.   
14                          MCLAIN FARMS, INC.   
17                         BONNIE PLANTS, INC.   
18                 B & W QUALITY GROWERS, INC.   
...                                        ...   
2700                  HARRAL LIVESTOCK CO. LLC   
2701                   ECOSYSTEM CONCEPTS INC.   
2702                        SIDDOWAY SHEEP CO.   
2705                    SATHER MANAGEMENT INC.   
2711                   C R KOEHL AND SONS INC.   

                                   name_clean  
4                      DUNSON HARVESTING, INC  
7     FARM LABOR ASSOCIATION FOR GROWERS, INC  
14                          MCLAIN FARMS, INC  
17                         BONNIE PLANTS, INC  
18                 B & W QUALITY GROWERS, INC  
...                                       ...  
2700                  HARRAL LIVESTOCK CO LLC  
2701                   ECOSYSTEM CONCEPTS INC  
2702                        SIDDOWAY SHEEP CO  
2705                    SATHER MANAGEMENT INC  
2711                   C R KOEHL AND SONS INC  

[576 rows x 2 columns]

Name                       name_clean
3              ANTON FERTILIZER INC.             ANTON FERTILIZER INC
4   GREAT PLAINS FLUID SERVICE, INC.  GREAT PLAINS FLUID SERVICE, INC
5                        PROMAX INC.                       PROMAX INC
13                     REIMER'S INC.                     REIMER'S INC
19                CISCO PRODUCE INC.                CISCO PRODUCE INC
23                     REIMER’S INC.                     REIMER’S INC
33                GERONIMO SHEEP CO.                GERONIMO SHEEP CO
46          ALTENDORF TRANSPORT INC.          ALTENDORF TRANSPORT INC
56                CISCO PRODUCE INC.                CISCO PRODUCE INC
72      SAXTONS RIVER ORCHARDS, INC.      SAXTONS RIVER ORCHARDS, INC
91          AVOYELLES HONEY CO., LLC          AVOYELLES HONEY CO, LLC

In [58]:
## your code here to print the head
## print
sub_jobs.head()
sub_debar.head()

CASE_NUMBER                                     CASE_STATUS  \
4   H-300-20231-774508            Determination Issued - Certification   
7   H-300-20233-780540  Determination Issued - Certification (Expired)   
14  H-300-20240-791807  Determination Issued - Certification (Expired)   
17  H-300-20244-795767            Determination Issued - Certification   
18  H-300-20245-799651            Determination Issued - Certification   

              RECEIVED_DATE            DECISION_DATE  \
4   2020-08-20 10:17:34.530  2020-10-01 00:00:00.000   
7   2020-09-04 14:26:00.270  2020-10-01 00:00:00.000   
14  2020-09-17 18:57:56.030  2020-10-01 00:00:00.000   
17  2020-09-16 12:58:58.600  2020-10-01 00:00:00.000   
18  2020-09-02 11:17:48.730  2020-10-01 00:00:00.000   

    TYPE_OF_EMPLOYER_APPLICATION H2A_LABOR_CONTRACTOR  \
4            Individual Employer                    Y   
7   Association - Joint Employer                    N   
14           Individual Employer                    N   
17           Individual Employer                    N   
18           Individual Employer                    N   

   NATURE_OF_TEMPORARY_NEED EMERGENCY_FILING  \
4                  Seasonal                N   
7                  Seasonal                N   
14                 Seasonal                N   
17                 Seasonal                N   
18                 Seasonal                N   

                               EMPLOYER_NAME           TRADE_NAME_DBA  ...  \
4                    DUNSON HARVESTING, INC.  Dunson Harvesting, Inc.  ...   
7   FARM LABOR ASSOCIATION FOR GROWERS, INC.                      NaN  ...   
14                        MCLAIN FARMS, INC.                      NaN  ...   
17                       BONNIE PLANTS, INC.                      NaN  ...   
18               B & W QUALITY GROWERS, INC.                      NaN  ...   

   TOTAL_HOUSING_RECORDS MEALS_PROVIDED MEALS_CHARGED  \
4                      8              N           NaN   
7                      1              Y          13.5   
14                     1              N           NaN   
17                     1              N           NaN   
18                    33              N           NaN   

   MEAL_REIMBURSEMENT_MINIMUM MEAL_REIMBURSEMENT_MAXIMUM PHONE_TO_APPLY  \
4                       12.68                       55.0    18632939888   
7                       13.50                       55.0    17605922256   
14                      12.68                       55.0    19125268436   
17                      12.68                       55.0    13343911328   
18                      12.68                       55.0    17725711135   

                   EMAIL_TO_APPLY       WEBSITE_TO_APPLY  \
4                             NaN  www.employflorida.com   
7                  flag@sfcos.com                    NaN   
14            kim@mclainfarms.com                    NaN   
17  dennis.ucles@bonnieplants.com                    NaN   
18                            NaN  www.employflorida.com   

    TOTAL_ADDENDUM_A_RECORDS                               name_clean  
4                          4                   DUNSON HARVESTING, INC  
7                          9  FARM LABOR ASSOCIATION FOR GROWERS, INC  
14                         1                        MCLAIN FARMS, INC  
17                         0                       BONNIE PLANTS, INC  
18                         7               B & W QUALITY GROWERS, INC  

[5 rows x 139 columns]

Name     City, State  \
3              ANTON FERTILIZER INC.     Dighton, KS   
4   GREAT PLAINS FLUID SERVICE, INC.  Greensburg, KS   
5                        PROMAX INC.  Whitewater, KS   
13                     REIMER'S INC.       Lakin, KS   
19                CISCO PRODUCE INC.       Cairo, GA   

                                    Violation Duration  Start date   End date  \
3   Failure to respond to audit (no response)  2 years   3/30/2014  3/29/2016   
4   Failure to respond to audit (no response)  2 years   3/30/2014  3/29/2016   
5                Failure to Hire U.S. workers  2 years   5/15/2014  5/14/2016   
13  Failure to respond to audit (no response)  2 years   8/23/2014  8/22/2016   
19  Failure to respond to audit (no response)  2 years  12/10/2014  12/9/2016   

    viol_num                       name_clean  
3        1.0             ANTON FERTILIZER INC  
4        1.0  GREAT PLAINS FLUID SERVICE, INC  
5        1.0                       PROMAX INC  
13       1.0                     REIMER'S INC  
19       1.0                CISCO PRODUCE INC

## 2.4 More joins and more cleaning (5 points)

A. Conduct another inner join between `jobs` and `debar_clean` now using the `name_clean` column; print the result. Did the cleaning result in any more employers matched between the two datasets?

B. Create a new column in `debar_clean` called `name_clean_2` that uses regex to take the following name in that dataset:

- `SLASH E.V. RANCH LLP` in the `debar_clean` dataset

And cleans it up so that it matches with this employer in `jobs`

- `SLASH EV RANCH` in the `jobs` dataset

Eg a pattern to remove the dots in the EV and the space+LLP-- you can apply the pattern to all employer names in debar_clean (so don't need to worry about only applying it to that one employer)


C. Conduct a left join using `name_clean_2` as the join column where the left hand dataframe is `jobs`; right hand dataframe is `debar_clean`, store the result as a dataframe, and print the rows where the merge indicator indicates the row was found in both dataframe

**Note**: this manual cleaning process is inefficient and helps motivate why talked about fuzzy matching. Fuzzy matching could recognize that Slash EV ranch is a highly similar string to slash ev ranch llp and match them without us needing to use regex to make the strings identical.

In [131]:
## your code here
merged_jobs2 = jobs.merge(debar_clean, on="name_clean", how = "inner")
merged_jobs2
# No more employers were matched in this merge, still just one.

regex2 = r"(\w+)(\.)(\w)(\.)(.+)(\sLLP)"

clean_name = [re.sub(regex2, r"\1\3\5", name) for name in debar_clean["name_clean"]]
debar_clean["name_clean_2"] = clean_name

merged_jobs3 = jobs.merge(debar_clean, left_on="name_clean", right_on="name_clean_2", how = "left", indicator="naics_merge_status")
merged_jobs3[merged_jobs3["naics_merge_status"] == "both"]



CASE_NUMBER                           CASE_STATUS  \
0  H-300-20287-876656  Determination Issued - Certification   

             RECEIVED_DATE            DECISION_DATE  \
0  2020-10-20 09:20:32.010  2020-11-09 00:00:00.000   

  TYPE_OF_EMPLOYER_APPLICATION H2A_LABOR_CONTRACTOR NATURE_OF_TEMPORARY_NEED  \
0          Individual Employer                    Y                 Seasonal   

  EMERGENCY_FILING   EMPLOYER_NAME TRADE_NAME_DBA  ... EMAIL_TO_APPLY  \
0                Y  RAFAEL BARAJAS            NaN  ...            NaN   

                WEBSITE_TO_APPLY TOTAL_ADDENDUM_A_RECORDS      name_clean  \
0  https://seasonaljobs.dol.gov/                        7  RAFAEL BARAJAS   

             Name       City, State                         Violation  \
0  RAFAEL BARAJAS  Sebring, Florida  Non-payment of certification fee   

   Duration  Start date   End date  
0    1 year   9/23/2016  9/22/2017  

[1 rows x 145 columns]

CASE_NUMBER                           CASE_STATUS  \
791   H-300-20287-876656  Determination Issued - Certification   
1115  H-300-20306-894148  Determination Issued - Certification   

                RECEIVED_DATE            DECISION_DATE  \
791   2020-10-20 09:20:32.010  2020-11-09 00:00:00.000   
1115  2020-11-02 18:11:29.140  2020-11-24 00:00:00.000   

     TYPE_OF_EMPLOYER_APPLICATION H2A_LABOR_CONTRACTOR  \
791           Individual Employer                    Y   
1115          Individual Employer                    N   

     NATURE_OF_TEMPORARY_NEED EMERGENCY_FILING   EMPLOYER_NAME TRADE_NAME_DBA  \
791                  Seasonal                Y  RAFAEL BARAJAS            NaN   
1115                 Seasonal                N  SLASH EV RANCH            NaN   

      ...    name_clean_x                  Name       City, State  \
791   ...  RAFAEL BARAJAS        RAFAEL BARAJAS  Sebring, Florida   
1115  ...  SLASH EV RANCH  SLASH E.V. RANCH LLP         Rifle, CO   

                             Violation Duration  Start date    End date  \
791   Non-payment of certification fee   1 year   9/23/2016   9/22/2017   
1115                     WHD Debarment   1 year  11/15/2014  11/14/2015   

              name_clean_y    name_clean_2  naics_merge_status  
791         RAFAEL BARAJAS  RAFAEL BARAJAS                both  
1115  SLASH E.V. RANCH LLP  SLASH EV RANCH                both  

[2 rows x 148 columns]

# 3. Optional extra credit 1: regex to separate companies from individuals (1 point)

You notice some employers in `debar_clean` have both the name of the company and the name of individual, e.g.:
    
COUNTY FAIR FARM (COMPANY) AND ANDREW WILLIAMSON (INDIVIDUAL)*

Use the uppercase/cleaned `name_clean` in `debar_clean`

A. Write a regex pattern that does the following:
    - Captures the pattern that occurs before COMPANY if (COMPANY) is in string; so in example above, extracts COUNTY FAIR FARM 
    - Captures the pattern that occurs before INDIVIDUAL if (INDIVIDUAL) is also in string -- so in above, extracts ANDREW WILLIAMSON (so omit the "and")
    
B. Test the pattern on `pos_example` and `neg_example`-- make sure former returns a list (if using find.all) or match object (if using re.search) with the company name and individual name separated out; make sure latter returns empty
    
**Hints and resources**: for step A, you can either use re.search, re.match, or re.findall; don't worry about matching B&R Harvesting and Paul Cruz (Individual)

- Same regex resources as above
    

In [ ]:
pos_example = "COUNTY FAIR FARM (COMPANY) AND ANDREW WILLIAMSON (INDIVIDUAL)*"
neg_example = "CISCO PRODUCE INC"

## your code here to define the pattern

## your code here to apply it to the pos_example

## your code here to apply it to the negative example

C. Iterate over the `name_clean` column in debar and use regex to create two new columns in `debar_clean`:
   - `co_name`: A column for company (full `name_clean` string if no match; pattern before COMPANY if one extracted)
   - `ind_name`: A column for individual (full `name_clean` string if no match; pattern before INDIVIDUAL if one extracted)
 


In [ ]:
# your code here

   
D. Print three columns for the rows in `debar_clean` containing the negative example and positive example described above (county fair farm and cisco produce):

- `name_clean`
- `co_name`
- `ind_name`
- `Violation`

**Note**: as shown in the outcome there may be duplicates of the same company reflecting different violations

In [ ]:
# your code here

# 4. Optional extra credit 2 (up to 3 points)

- For 1 point extra credit, create a visualization with 1+ of the existing fields in either the raw `jobs` or `debar` data. We'll be showing cool visualizations in class so use your imagination! Options could include visualizing between-state or over-time variation

- For 3 points extra credit instead, geocode the employer addresses in `jobs` and plot the addresses of jobs as points overlaid on top of a map of Georgia 
    - **Note**: this extra credit involves Googling since we have not yet covered spatial data. 
        - For discussion of how to geocode addresses -> lat/long, see: https://www.natasshaselvaraj.com/a-step-by-step-guide-on-geocoding-in-python/ 
        - For discussion of plotting lat/long dots against a map, see this discussion of geopandas: https://towardsdatascience.com/plotting-maps-with-geopandas-428c97295a73
    - Relevant columns include `EMPLOYER_ADDRESS_1` 
    - The geocoding might have a long runtime so feel free to implement it in a separate .py script that you submit alongside your notebook and to just read in the geocoded data

In [ ]:
## your code here